### PR TITLE
Rutger wip

### PR DESCRIPTION
**Pull request by parutger**
Reintroduces String-based searching for Accounts
Converts console logger to Reporter
Moves missed-block out of main and into chain_monitor

Renames eslint files 
Implements most feedback from polkadot-basic-notification/pull/1
This PR is currently not backwards compatible due to config file format and filter notation.

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -1,18 +0,0 @@
-{
```
Comment (kianenigma) [inline .eslintrc:1]: why remove this?

Comment (kianenigma) [reply .eslintrc:1]: in fact, if you being it back and ask your editor to format the files, everything will look much nicer

Comment (kianenigma) [inline .eslintrc:1]: why is this being replaced with a smaller version?

File: pm2.config.js [hunk 2]
```diff
@@ -1,13 +0,0 @@
-module.exports = {
```
Comment (kianenigma) [inline pm2.config.js:1]: can stay.

File: src/chain_monitor.ts [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+import "@polkadot/api-augment";
+import "@polkadot/types-augment";
+
+const DEBUG = false;
+
+export interface ExtrinsicItem {
+    index: number,
+    section: string,
+    method: string,
+    account: ExtendedAccount,
```
Comment (kianenigma) [inline src/chain_monitor.ts]: who is account? what if it is an unsigned extrinsic?

File: src/chain_monitor.ts [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+export interface EventItem {
+    section: string,
+    method: string,
+    data: string,
+}
+
+// This holds the raw blockdata,
+// the blockhandler extracts relevant data into a "Report" item.
+export interface blockData {
```
Comment (kianenigma) [inline src/chain_monitor.ts]: Interfaces should be named capital camel case.

File: src/chain_monitor.ts [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+
+export class ChainMonitor {
+    private provider!: WsProvider;
+    private api!: ApiPromise;
+    chain!: string;
+
+
+    constructor(
```
Comment (kianenigma) [inline src/chain_monitor.ts:54]: I don't know what formetter you use here (or none), but I don't agree with it :D

File: src/config.ts [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+export interface ExtendedAccount {
+    address: Address,
+    label: string,
+}
+
+export interface AppConfig {
+    endpoints: string[],
+    accounts: ExtendedAccount[],
+    eventFilter: "all" | string[],
```
Comment (kianenigma) [inline src/config.ts]: what if I want to listen to transactions?

File: src/config.ts [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    matrix?: MatrixConfig,
```
Comment (kianenigma) [inline src/config.ts]:
```
reporters: {
  email: ..,
  matrix: ..,
  console: ..,
}
```

is better since all of them are nested in `reporters`. For example. how can your app now log how many reporters it has? it can't.

File: src/config.ts [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+            this.config.eventFilter = "all"
+        }
+        // Matrix
+        // Environment variable overwrites config entry
+        if (typeof process.env.MATRIX_TOKEN !== undefined && this.config.matrix !== undefined) {
+            this.config.matrix.accessToken = process.env.MATRIX_TOKEN || this.config.matrix.accessToken;
+        }
+    }
+
+    getReporters() {
```
Comment (kianenigma) [inline src/config.ts]:
```suggestion
    getReporters(): Reporters[] {
```

Comment (kianenigma) [inline src/config.ts]: if you would have not changed my structure for the json file, as noted in https://github.com/paritytech/polkadot-basic-notification/pull/2/files#r906646198, you wouldn't even need this function. It would simply be `appConfig.reporters`

File: src/config.ts [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+        // EventFilter
+        if (this.config.eventFilter == undefined) {
+            console.warn("No 'eventFilter' section found in config, defaulting to 'all'");
+            this.config.eventFilter = "all"
+        }
+        // Matrix
+        // Environment variable overwrites config entry
+        if (typeof process.env.MATRIX_TOKEN !== undefined && this.config.matrix !== undefined) {
+            this.config.matrix.accessToken = process.env.MATRIX_TOKEN || this.config.matrix.accessToken;
```
Comment (kianenigma) [inline src/config.ts]: since you are keen on adding features, using `dovenv` would be a good improvement.

File: src/health_probes.ts [hunk 10]
```diff
@@ -0,0 +1,32 @@
+import { createServer, Server } from 'http';
+
+/* 
+ * This is a very simplified readiness probe that responds to any url with a status code
+ * Can be further extended to provide metrics and 
+ */
+
+export class Healthprobe {
+
```
Comment (kianenigma) [inline src/health_probes.ts]:
```suggestion
```

Comment (kianenigma) [reply src/health_probes.ts]: you generally have a lot of newlines that could go away and make the code look a bit scattered. Setting up a strict formatter will help.

File: src/index.ts [hunk 11]
```diff
@@ -1,279 +1,122 @@
...
+    readiness.listen();
+
+    const configurator = new Config()
+    const config: AppConfig = configurator.config;
+    const reporters: Reporter[] = configurator.getReporters();
+
+    // This function is passed to chainMonitor.subscribeHandler and is called every block.
+    // Here is the main application logic that determines if a report is created
+    // and which extrinsics/events are included
+    const BlockHandler = async (blockHeader: Header, chainMonitor: ChainMonitor) => {
```
Comment (kianenigma) [inline src/index.ts]: Would make the main code much easier to read if you make this a function rather than a const.

File: src/config.ts (hunk 9 above)
Comment (parutger) [reply src/config.ts]: Let's do that in at a later time

File: src/chain_monitor.ts (hunk 3 above)
Comment (parutger) [reply src/chain_monitor.ts]: Unsigned extrinsics now take either the related account from the extrinsic as a string, OR remain undefined if there are no accounts set and the extrinsic is in the extrinsicFilter.

File: src/config.ts (hunk 6 above)
Comment (parutger) [reply src/config.ts]: You can now use the extrinsicsFilter to listen to specific methods.

File: src/config.ts (hunk 7 above)
Comment (parutger) [reply src/config.ts]: Agreed, done

File: .eslintignore [hunk 12]
```diff
@@ -1,2 +0,0 @@
-node_modules
```
Comment (kianenigma) [inline .eslintignore:1]: can probably also come back as well, although I am not 100% sure if it makes any difference 🤷

File: src/logger.ts [hunk 13]
```diff
@@ -1,18 +0,0 @@
-import * as winston from 'winston'
```
Comment (kianenigma) [inline src/logger.ts:1]: Why is this removed?

File: src/reporters/index.ts [hunk 14]
```diff
@@ -1,228 +1,52 @@
...
+export { EmailReporter } from "./email";
+export { ConsoleReporter } from "./console";
+export type { Report } from "../chain_monitor";
 
 export interface Reporter {
-	report(report: Report): Promise<void>;
-
-}
-
-export class GenericReporter  {
```
Comment (kianenigma) [inline src/reporters/index.ts:47]: Why is this removed, and replaced with a function? 

Again, you are removing useful programming abstractions. If your goal is simplicity, I disagree that this helps. 

Now you are hard-coding things like "how to get a subscan link" inside of the `ReportHTML`. What if in a month, we need to add a markdown exporter as well? should we hard-code this in there as well? 

This is exactly why I abstracted this in `GenericReporter`.

**Review by kianenigma — CHANGES_REQUESTED** [review]
I am seeing things like https://github.com/paritytech/polkadot-basic-notification/pull/2/files#r912042262 too many times here and I feel like we might be losing some time.

@parutger I gladly want you to contribute to this repo and take it over, but it is very difficult for me to sign-off on code changes that based on my experience are mostly wrong and a step backwards. 

I suggest: let's start with a smaller PR. Let's first only add your fix for the skipped finalized blocks, and then we will figure out how to take this forward.

In the meantime, let's deploy a few instances of this as well, and see what features we might be missing.

File: src/index.ts [hunk 15]
```diff
@@ -1,279 +1,116 @@
...
+                data: JSON.stringify(extrinsic.toHuman()),
+            };
+
+            // Add to report
+            extrinsicItems.push(item);
+        });
+
+        // For each Event
+        const eventItems: EventItem[] = [];
+        for (const event of data.events) {
```
Comment (kianenigma) [inline src/index.ts:62]: Events are not checked against account anymore.
